### PR TITLE
Fix issues identified with the Hapi 17 update

### PR DESCRIPTION
**Pull request by jhsimms**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/hapijs/hapi/blob/master/API.md
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by jhsimms** [comment]
@SimonSchick This hopefully addresses all of the issues you reported.  I wasn't very organized in going through your notes, so it's possible I missed one.  Please just point out if you do.

I tested this against my local Hapi server, but we don't make extensive use of features, so it will be better for you to test some of these against your implementation to see how it fares.

Finally, we discussed collapsing these all into a single file.  I held off on that to make review of this simpler.  However, if we can hold off on a merge, I figure we can go back and forth on the PR as it exists now, then, once everyone is content, I can collapse all the files into a single file prior to merging.

CC @rafaelsouzaf

File: types/hapi/definitions/util/lifecycle.d.ts
Comment (SimonSchick) [inline types/hapi/definitions/util/lifecycle.d.ts]: You make make that a type now.

File: types/hapi/definitions/server/server-ext.d.ts
Comment (SimonSchick) [inline types/hapi/definitions/server/server-ext.d.ts]: Do these narrow correctly with `type`?

**Comment by SimonSchick** [comment]
I think the narrowing of `ext` on routes still isn't perfect but all of these fixes are very good! Could use a few test cases tho.

File: types/hapi/definitions/server/server-events.d.ts
Comment (SimonSchick) [inline types/hapi/definitions/server/server-events.d.ts]: I don't think we need the generic version, wouldn't this still permit typos to be uncaught?

File: types/hapi/definitions/server/server-events.d.ts
Comment (SimonSchick) [inline types/hapi/definitions/server/server-events.d.ts]: See below.

File: types/hapi/definitions/server/server-events.d.ts
Comment (SimonSchick) [inline types/hapi/definitions/server/server-events.d.ts]: Aren't these the same as above?

Also you could get a little bit _smart_ to avoid writing the same code over and over again, which saves code (tradeoff is weird autocompletion is VSCode).

https://github.com/SimonSchick/EliteDangerousUtils/blob/master/src/EDLog/EDLog.ts#L307

On the nice side you can use this as a generic interface and declare all methods at once!

**Review by SimonSchick — CHANGES_REQUESTED** [review]
Yea, so we need to get rid of the generic _catch all_ overloads, these permit bad usage.

**Comment by SimonSchick** [comment]
Hey @jhsimms would you mind giving me write perms on your DT fork? We could speed things up a little that way.

**Comment by jhsimms** [comment]
Yes @SimonSchick, done.

**Comment by SimonSchick** [comment]
@jhsimms is there any IM I can potentially talk to you on?

**Comment by SimonSchick** [comment]
@jhsimms I performed some changes that I think will be correct in the long run.
Please ping me at any time if you feel any of these are unfit or bad.

**Comment by SimonSchick** [comment]
We will have to update some hapi dependent packages it seems, many authors did rely on using `config` instead of `options`, this should have relatively little impact.

`Vision` seems to have broken tests in general, as it's typings do not allow passing `engines`, I fixed that.

**Comment by SimonSchick** [comment]
Ok, I instead decided to remove the `vision` dependency, I don't know why hapi's tests relied on a dependant of itself, kind of a bad idea in the first place, this should not destabilize typings in any way however.

**Comment by SimonSchick** [comment]
Needless to say this needs to be squash(merged/ed before merge).

**Comment by jhsimms** [comment]
So I don't forget tomorrow, I just realized that the server.dependency function is not typed to be async.  It should be:

`dependency(dependencies: string | string[], after?: ((server: Server) => Promise<void>)): void;`

**Comment by SimonSchick** [comment]
Go on and fix it, I'm on mobile rn and can't really work on this.

**Review by SimonSchick — APPROVED** [review]

**Comment by lenovouser** [comment]
The `request.response` object is missing some properties such as `.isBoom` and `.output`

 - [hapijs.com/api#errors](https://hapijs.com/api#errors)
 - [hapijs/hapi/lib/request.js#L390](https://github.com/hapijs/hapi/blob/304074a4f0067b8de7674be059fbfbf21103b8b6/lib/request.js#L390)

basically you can e.g. to this in a hapi extension:

```ts
server.ext('onPreResponse', (request: Request, h: ResponseToolkit) => {
    const response = request.response.isBoom ? request.response.output : request.response;
}
```

to check if the request already failed somewhere previous in the extension chain. If so, the `response.output` includes the modified erroneous response for some reason - it is kind of documented in [hapijs.com/api#errors](https://hapijs.com/api#errors) and some extensions use it too:

 - [nlf/blankie/lib/index.js#L246](https://github.com/nlf/blankie/blob/master/lib/index.js#L246)
 - [GitHub Code Search](https://github.com/search?l=JavaScript&q=%22request.response.output%22+hapi&type=Code&utf8=%E2%9C%93)

**Comment by jhsimms** [comment]
@lenovouser I added the props.  I replicated what was in the v16 typings.

**Comment by SimonSchick** [comment]
@jhsimms I do believe the object doesn't have the boom properties but it actually is an instance of `Boom`, `isBoom` doesn't appear to be assigned anywhere inside of `hapi` core so it must be a boom instance.

**Comment by jhsimms** [comment]
Hmm, thanks @SimonSchick.  I'm on the same page.  I was thinking it was strange, though the examples do show it.  I also don't think there's a great way to type it.

As for the typings, I suppose we should take isBoom and the other properties out, and people that need that functionality can use a union type between a hapi response and a boom object to determine what they have.  Does that work for you?

**Comment by SimonSchick** [comment]
Yea, just make it a type union and we're good!

**Comment by jhsimms** [comment]
@SimonSchick Done, I'm going to collapse these all into a single file now.  Do you see any more changes needed prior to doing that?

**Comment by SimonSchick** [comment]
Not at this moment, if I find any more issues I will bring them up.

**Comment by corbinu** [comment]
@jhsimms Can you merge in my PR here also? https://github.com/DefinitelyTyped/DefinitelyTyped/pull/23632

**Comment by jhsimms** [comment]
@corbinu Done

File: types/hapi/index.d.ts
Comment (lenovouser) [inline types/hapi/index.d.ts]: Can also be typed 

```ts
import Podium from "podium";
```

as far as I know

Comment (corbinu) [reply types/hapi/index.d.ts]: No it can't this is the change I asked jhsimms to make: https://github.com/Microsoft/TypeScript/issues/21872

Comment (corbinu) [reply types/hapi/index.d.ts]: Sorry your correct that's my bad I think I have been up to long my eyes are playing tricks on me. Either is fine.

Comment (jhsimms) [reply types/hapi/index.d.ts]: The linter and the compiler are both angry if I try to change it from the above.  I left it as it, as there are a lot of other issues (linter) that arose from collapsing into a single file that I needed to address.

**Comment by jhsimms** [comment]
@SimonSchick @corbinu @lenovouser I collapsed into a single file and fixed all the dependent packages.  The tests are passing (for me), but I still need to test this out against my server.  I'd appreciate if you could do the same.

I rebased onto the upstream master prior to collapsing.  I have not squashed yet, for your convenience in inspecting the changes.

I did a single commit that was simply collapsing everything, as is, into a single file, excluding unnecessary imports.  That caused one big problem, however.  Apparently, the DT linter was not checking any of the files under the definitions directory.  As soon as I collapsed it, the linter reported several hundred problems.  I addressed those in a 2nd commit.

**Comment by SimonSchick** [comment]
@jhsimms I will be OOF for a couple of days, I'd ask you to merge this for now as I cannot test this with our system for now.

**Comment by jhsimms** [comment]
Thanks, I'll test it out and then update this ticket when it's verified as ready to merge.

**Comment by jhsimms** [comment]
The recent changes to the decorate broke my server, primarily because it no longer allows args to be passed to the decorated method.  The extend functionality was also missing.

**Comment by jhsimms** [comment]
Ok, this is squashed and ready to go, from my perspective.  Please let me know if it looks like any other changes are needed.

**Comment by corbinu** [comment]
@jhsimms I tried it however I am not getting an error for my routes as they all have a config property. Seems like this was removed from ServerRoute.

```
Object literal may only specify known properties, and '"config"' does not exist in type 'ServerRoute | ServerRoute[]'
```

File: types/hapi/index.d.ts
Comment (lenovouser) [inline types/hapi/index.d.ts]: I think you still need to make a union with `Boom` here

**Review by lenovouser — CHANGES_REQUESTED** [review]

File: types/hapi/index.d.ts
Comment (SimonSchick) [reply types/hapi/index.d.ts]: That’d be on the response property.

**Comment by SimonSchick** [comment]
@corbinu config is deprecated, we could add it back but I don’t see much value in it.

File: types/hapi/index.d.ts
Comment (lenovouser) [reply types/hapi/index.d.ts]: Ah, exactly. My mistake.

Comment (SimonSchick) [reply types/hapi/index.d.ts]: Can you approve it then? 😛

File: types/hapi/index.d.ts
Comment (lenovouser) [inline types/hapi/index.d.ts]: @SimonSchick I just commented on the wrong line, it still needs to be done.

**Review by lenovouser — CHANGES_REQUESTED** [review]

File: types/hapi/index.d.ts
Comment (SimonSchick) [reply types/hapi/index.d.ts]: Oh sorry, I thought that was done already.

Comment (SimonSchick) [reply types/hapi/index.d.ts]: I might actually be able to do a commit tonight despite my previous comment.

**Comment by corbinu** [comment]
@SimonSchick Strange there are still examples of it in hapi's api docs. 

Ok no worries then the irony is I am moving off hapi on the 1st anyway. Just was trying to get the podium thing fixed in the meantime.

**Comment by SimonSchick** [comment]
@corbinu the docs have not been correctly updated in many places.

File: types/hapi/index.d.ts
Comment (jhsimms) [reply types/hapi/index.d.ts]: I overlooked this, but it will fail to compile as soon as I included Boom.BoomError.  For instance:

```
export interface ResponseObject extends Podium, Boom.BoomError {
```

However, there's a type conflict with `response.message`.  ResponseObject defines it as a function.  Boom defines `message` as a string.  The only way it will compile is if I change it to `message: any`, which I don't really want to do.  Is there another way of handling this?

Comment (SimonSchick) [reply types/hapi/index.d.ts]: No, the `response` property on request is `RequestObject | Boom.BoomError`.

**Review by lenovouser — APPROVED** [review]

**Comment by SimonSchick** [comment]
I went ahead and verified it's actually a boom error object, this is correct now.

**Comment by SimonSchick** [comment]
Code:
```
> void(h = require('hapi'))
undefined
> void(b = require('boom'))
undefined
> void(s = new h.Server())
undefined
> void(s.ext('onPreResponse', (req, h) => { console.log(req.response); return h.continue; }))
undefined
> void(s.route({ path: '/', method: 'get', handler(req, tk) { throw b.badRequest(); } }))
undefined
> void(s.inject({ url: '/' }))
undefined
> { Error: Bad Request
    at handler (repl:1:69)
    at module.exports.internals.Manager.execute (/Users/v-sischi/Projects/misc/vendor/DefinitelyTyped/node_modules/hapi/lib/toolkit.js:35:106)
    at Object.internals.handler (/Users/v-sischi/Projects/misc/vendor/DefinitelyTyped/node_modules/hapi/lib/handler.js:52:48)
    at exports.execute (/Users/v-sischi/Projects/misc/vendor/DefinitelyTyped/node_modules/hapi/lib/handler.js:37:36)
    at module.exports.internals.Request._lifecycle (/Users/v-sischi/Projects/misc/vendor/DefinitelyTyped/node_modules/hapi/lib/request.js:261:62)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:160:7)
  data: null,
  isBoom: true,
  isServer: false,
  output:
   { statusCode: 400,
     payload: { statusCode: 400, error: 'Bad Request', message: 'Bad Request' },
     headers: {} },
  reformat: [Function],
  message: 'Bad Request',
  typeof: [Function: badRequest] }
```

**Comment by lenovouser** [comment]
LGTM! 🚀

**Comment by SimonSchick** [comment]
@andy-ms can we have this merged? 😄

**Comment by lenovouser** [comment]
How often do the types get published? I thought this was done via webhook?

**Pull request merged**
